### DISCUSSION TITLE
关于raft协议voteFor元数据的用途

### DISCUSSION THREAD

**Discussion by FantasyChen-DH** [sofastack/sofa-jraft#950] [Q&A] [answered]
对于voteFor这一元数据的用途我有些疑问。原本来说voteFor这一参数是用于防止重复投票的情况发生，但是在我编码过程中，却发现该参数好像没有使用的必要。具体分析如下：
有两个节点分别为A和B，A向B发起了拉票请求。接下来有以下三种情况：

1. termB > termA，B节点应当拒绝此次拉票。
2. termB == termA，如果B是leader，则证明该term已经存在了一个leader，A就不应该当选为leader，所以拒绝；如果B是candidate，则代表在该任期B投票给了自己，所以拒绝；如果B是follower，B会存在两种情况：已经投票给了一个Candidate并更新了任期或者B跟随一个leader，两种情况也都不应该投票给A节点。所以B节点应当拒绝此次拉票。
3. termB < termA，这种情况无论voteFor是否为空，对于B节点来说都已经进入了一个新的任期，所以只要节点A的日志比B新，B就应该投票给节点A。

综上所述，我使用不到voteFor这一参数。所以感到十分疑惑，感谢解答！

**Comment by shihuili1218** [comment]
可能少考虑了一种情况呢，就是成员宕机立马重启。比如：
abc的term为1，a发起选举递增term，ab投票给a，term更新为2。
b宕机立马重启。
c发起选举递增term为2，如果没有votefor，b会再投票给c。

代码如下：这里判断的是大于等于，2的term是可以通过多次的（以b为例，a的term=2，c的term=2，都能通过这个逻辑）。
https://github.com/sofastack/sofa-jraft/blob/fdf737850097eb1bf20d34b8caaa0bb6f0c92ad7/jraft-core/src/main/java/com/alipay/sofa/jraft/core/NodeImpl.java#L1781

那能不能把大于等于改成大于？
也是不能的，因为正常情况下，term 可以相等，logid大能竞选成功

**Reply by FantasyChen-DH** [reply]
嗯 说的很有道理 但是我认为term相同是无需要投票的 就如我上面论述一样 我找不到term相同时需要投票的情况 所以才会有疑问

**Reply by shihuili1218** [reply]
> 嗯 说的很有道理 但是我认为term相同是无需要投票的 就如我上面论述一样 我找不到term相同时需要投票的情况 所以才会有疑问

成员之间的term相同，这是很常见的场景呀。leader选举，先比对term，term一样再比对logid

**Reply by FantasyChen-DH** [reply]
我的意思是term相同时需要给节点投票的情况我不太明白。。。 在我的理解中term相同时应该不需要投票了，就像我之前论述的一样，相同时都应该拒绝（虽然论文中是大于等于都进行投票）可以举一点term相同时还要投票给那个节点的例子吗。。

**Reply by FantasyChen-DH** [reply]
额 看了案例我还是不太明白。。可能是我的问题不够清晰
https://github.com/sofastack/sofa-jraft/blob/fdf737850097eb1bf20d34b8caaa0bb6f0c92ad7/jraft-core/src/main/java/com/alipay/sofa/jraft/core/NodeImpl.java#L1781
我的问题是，为什么这一行不能改成大于呢？因为任期相同时的拉票情况也就是termB == termA的情况，我在上面已经讨论了一遍。。找不到必须投票的理由。。/(ㄒoㄒ)/~~

**Reply by shihuili1218** [reply, marked as answer]
考虑了很多场景，大于等于改成大于，应该是不影响正确性。

但是加上等于号，有一种情况可以加快选举进程

A term=2， index=201
B term=2， index=203
C term=2， index=203


A发起选举，term3，线程暂停，还未投票给A
B发起选举，term3，
A收到B的选举消息：
- 如果没有等于号，A不会投票给B，而A线程恢复过来，投票给自己，A也是不能晋升的,因为index小。
- 如果有等于的存在，A会投票给B，加快B的选举

**Reply by FantasyChen-DH** [reply]
好的 感谢解答！

**Comment by killme2008** [comment]
@shihuili1218 回答的很好了。本质上就是为了加速选举，尽快收敛。

> 已经投票给了一个Candidate并更新了任期或者B跟随一个leader，两种情况也都不应该投票给A节点。所以B节点应当拒绝此次拉票。

已经投了一个 Candidate， votedId 就是记录投了谁，用来拒绝。

https://github.com/sofastack/sofa-jraft/blob/fdf737850097eb1bf20d34b8caaa0bb6f0c92ad7/jraft-core/src/main/java/com/alipay/sofa/jraft/core/NodeImpl.java#L1821
